tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,158 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow.op;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * A class to manage scoped (hierarchical) names for operators.
+ *
+ * <p>{@code NameScope} manages hierarchical names where each component in the hierarchy is
+ * separated by a forward slash {@code '/'}. For instance, {@code nn/Const_72} or {@code
+ * nn/gradient/assign/init}. Each scope is a subtree in this hierarchy.
+ *
+ * <p>Use {@code NameScope} to group related operations within a hierarchy, which for example lets
+ * tensorboard coalesce nodes for better graph visualizations.
+ *
+ * <p>This class is package private, user code creates {@link Scope} which internally delegates
+ * calls to an underlying {@code NameScope}.
+ *
+ * <p>This class is thread-safe.
+ */
+final class NameScope {
+
+  NameScope withSubScope(String scopeName) {
+    if (baseName == null) {
+      checkPattern(OP_NAME_REGEX, scopeName);
+    } else {
+      checkPattern(SUBSCOPE_NAME_REGEX, scopeName);
+    }
+
+    if (baseOpName != null) {
+      // Use the base name instead to derive the subscope.
+      scopeName = baseOpName;
+    }
+
+    String newBaseName = fullyQualify(makeUnique(scopeName));
+    return new NameScope(newBaseName, null, null);
+  }
+
+  NameScope withName(String name) {
+    checkPattern(OP_NAME_REGEX, name);
+
+    // All context except for the baseOpName is shared with the new scope.
+    return new NameScope(baseName, name, ids);
+  }
+
+  String makeOpName(String opName) {
+    checkPattern(OP_NAME_REGEX, opName);
+
+    if (baseOpName != null) {
+      opName = baseOpName;
+    }
+    opName = makeUnique(opName);
+    return fullyQualify(opName);
+  }
+
+  /**
+   * Create a new, root-level namescope.
+   *
+   * <p>A root-level namescope generates operator names with no components, like {@code Const_72}
+   * and {@code result}.
+   */
+  NameScope() {
+    this(null, null, null);
+  }
+
+  private NameScope(String baseName, String baseOpName, Map<String, Integer> ids) {
+    this.baseName = baseName;
+    this.baseOpName = baseOpName;
+    if (ids != null) {
+      this.ids = ids;
+    } else {
+      this.ids = new HashMap<String, Integer>();
```

Comment:
I just realized that if we want to allow users to name ops with slashes in them (as is possible in Python), then we'd probably need the name-map attached to the root scope and not create new hash-maps per sub-scope - i.e., a single map that covers all the nodes in the graph. I believe that is the case in the C++ scope as well.

Should we replicate that here?